File: src/main/java/frc/robot/vision/VisionBackend.java
package frc.robot.vision;

import java.util.Optional;

import edu.wpi.first.math.Matrix;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.numbers.N1;
import edu.wpi.first.math.numbers.N3;

public abstract class VisionBackend {
    public abstract Optional<Measurement> getMeasurement();

    public class Measurement {
        public double timestamp;
        public Pose3d pose;
        public Matrix<N3, N1> stdDeviation;

        public Measurement(double timestamp, Pose3d pose, Matrix<N3, N1> stdDeviation) {
            this.timestamp = timestamp;
            this.pose = pose;
            this.stdDeviation = stdDeviation;
        }
    }
}

File: src/main/java/frc/robot/poseestimation/VisionBackend.kt
package frc.robot.poseestimation

import edu.wpi.first.math.Matrix
import edu.wpi.first.math.geometry.Pose3d
import edu.wpi.first.math.numbers.N1
import edu.wpi.first.math.numbers.N3
import java.util.*

abstract class VisionBackend {
    abstract val measurement: Optional<Measurement?>

    inner class Measurement(var timestamp: Double, var pose: Pose3d, var stdDeviation: Matrix<N3, N1>, var ambiguity: Double)
}